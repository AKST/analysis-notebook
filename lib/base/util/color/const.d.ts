export const COLOR = {
  RED: number,
  YELLOW: number,
  CYAN: number,
  LIME: number,
  MAGENTA: number,
  WHITE: number,
};
