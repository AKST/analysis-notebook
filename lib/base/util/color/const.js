export const COLOR = {
  RED: 0xff0000,
  YELLOW: 0xffff00,
  CYAN: 0x00ffff,
  LIME: 0x44ff00,
  MAGENTA: 0xff00ff,
  WHITE: 0xffffff,
};

