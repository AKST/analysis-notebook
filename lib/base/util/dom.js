/**
 * @param {Node} parent
 * @param {Node} child
 */
export function insertAtStart(parent, child) {
  if (parent.firstChild) {
    parent.insertBefore(child, parent.firstChild);
  } else {
    parent.appendChild(child);
  }
}

/**
 * @param {HTMLElement | undefined} element
 * @param {number | boolean | string | null | undefined} value
 */
export function setDisable(element, value) {
  if (element == null) return;
  switch (value) {
    case undefined:
    case null:
    case '':
    case false:
    case 0:
      element.removeAttribute('disabled');
      break;
    default:
      element.setAttribute('disabled', 'disabled');
  }
}
