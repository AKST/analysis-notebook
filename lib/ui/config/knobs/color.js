/**
 * @import { ColorKnobConfig } from '@base/runtime/config/type.ts';
 * @import { ConfigKnobLabel } from '../common/label.js';
 * @import { KnobElement } from '../type.ts';
 */

import { ChangeKnobEvent } from '../events.js';
import { generateCSS, updateOn } from '@base/dom_ui/index.js';
import { ColorInput as ColorInputComponent } from '../../inputs/color.js';
import '../common/label.js';

function getStyles() {
  return generateCSS({
    ':host': {
      gridColumn: 'span 4'
    },
    '.container': {
      display: 'contents'
    },
    '.color-input-container': {
      gridColumn: 'span 9',
    },
  });
}

/**
 * @implements {KnobElement}
 */
export class ColorKnob extends HTMLElement {
  gridColumn = 4;

  #value = 0xff0000; // Default red
  /** @type {ColorKnobConfig} */ #config;
  /** @type {ColorInputComponent | null} */ #colorInput = null;

  /**
   * @param {ColorKnobConfig} cfg
   * @param {ConfigKnobLabel | undefined} label
   */
  constructor(cfg, label, value = 0xff0000) {
    super();
    this.attachShadow({ mode: 'open' });
    this.label = label;
    this.#value = value;
    this.#config = cfg;
  }

  setup() {
    this.#colorInput = new ColorInputComponent(this.#value);
    this.#colorInput.className = 'color-input-container';
  }

  setEnable() {
    throw new Error('setEnabled is not implemented on ComplexKnob');
  }

  connectedCallback() {
    this.render();
    this.shadowRoot?.addEventListener('change', (e) => {
      if (e.target === this.#colorInput) {
        this.#value = /** @type {any} */ (e).detail.value;
        this.dispatchEvent(new ChangeKnobEvent(
          { value: this.#value, child: [] },
          { bubbles: true }
        ));
      }
    });
  }

  render() {
    updateOn(this.shadowRoot, [
      ['style', {}, [getStyles()]],
      ['div', { class: 'container' }, [
        this.label,
        this.#colorInput
      ]]
    ]);
  }
}

customElements.define('config-color-knob', ColorKnob);
