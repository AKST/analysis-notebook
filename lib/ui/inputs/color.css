:host {
  display: block;
}

.color-input {
  display: block;
  width: 100%;
  height: 24px;
  outline: none;
  background: transparent;
  cursor: pointer;
  border-radius: 0;
  padding: 0;
  border: var(--border-outset-grey);

  &::-webkit-color-swatch {
    border: none;
  }
}
