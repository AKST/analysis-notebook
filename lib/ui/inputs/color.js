import { generateCSS, updateOn } from '../../base/dom_ui/index.js';
import { colorHex, hexToColor } from '../../util/color/index.js';

function getStyles() {
  return generateCSS({
    ':host': {
      display: 'block',
    },
    '.color-input': {
      display: 'block',
      width: '100%',
      height: 24,
      outline: 'none',
      background: 'transparent',
      cursor: 'pointer',
      borderRadius: 0,
      padding: 0,
      border: 'var(--border-outset-grey)',
      '&::-webkit-color-swatch': {
        border: 'none',
      },
    },
  });
}

/**
 * Color Input Web Component
 * Based on RFC 006 - reusable input for use in knobs and other components
 */
export class ColorInput extends HTMLElement {
  static observedAttributes = ['value', 'disabled'];

  #value = 0xff0000; // Default red

  /**
   * @param {number} value
   */
  constructor(value) {
    super();
    this.#value = value
    this.attachShadow({ mode: 'open' });
  }

  connectedCallback() {
    this.render();
    this.setupEventListeners();
  }

  setupEventListeners() {
    this.shadowRoot?.addEventListener('change', (e) => {
      if (e.target instanceof HTMLInputElement) {
        this.#value = hexToColor(e.target.value);
        this.updateInput()
        this.dispatchEvent(new CustomEvent('change', {
          detail: { value: this.#value },
          bubbles: true
        }));
      }
    });
  }

  /**
   * @param {number} value
   */
  setValue(value) {
    this.#value = value;
    this.updateInput();
  }

  getValue() {
    return this.#value;
  }

  updateInput() {
    const input = this.#input;
    if (input) {
      input.value = colorHex(this.#value);
      input.style.background = colorHex(this.#value)
    }
  }

  render() {
    updateOn(this.shadowRoot, [
      ['style', {}, [getStyles()]],
      ['input', {
        type: 'color',
        class: 'color-input',
        styles: {
          backgroundColor: colorHex(this.#value),
        },
        value: colorHex(this.#value),
        disabled: this.hasAttribute('disabled') || null,
      }, []]
    ]);
  }

  /**
   * @returns {HTMLInputElement | null}
   */
  get #input () {
    const colorInput = this.shadowRoot?.querySelector('.color-input');
    return colorInput ? /** @type {HTMLInputElement} */ (colorInput) : null;
  }
}

customElements.define('color-input', ColorInput);
