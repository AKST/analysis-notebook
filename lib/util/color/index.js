export { COLOR } from './const.js';

/**
 * Convert hex color number to CSS hex string
 * @param {number} hex - Color as hex number (e.g., 0xff0000)
 * @returns {string} CSS hex string (e.g., "#ff0000")
 */
export function colorHex(hex) {
  const hexStr = hex.toString(16).padStart(6, '0');
  return `#${hexStr}`;
}

/**
 * Convert CSS hex string to hex color number
 * @param {string} hexString - CSS hex string (e.g., "#ff0000")
 * @returns {number} Color as hex number (e.g., 0xff0000)
 */
export function hexToColor(hexString) {
  const hex = hexString.replace('#', '');
  return parseInt(hex, 16);
}
