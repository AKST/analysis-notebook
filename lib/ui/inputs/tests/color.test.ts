import { expect, describe, it, beforeEach, vi } from 'vitest';
import userEvent from '@testing-library/user-event';
import { getByDisplayValue } from '@testing-library/dom';
import { ColorInput } from '../color.js';

describe('ColorInput', () => {
  let user: ReturnType<typeof userEvent.setup>;

  beforeEach(() => {
    document.body.innerHTML = '';
    user = userEvent.setup();
  });

  const create = (color: number) => {
    const colorInput = new ColorInput(color);
    document.body.appendChild(colorInput);
    return colorInput;
  };

  it('renders color input correctly', () => {
    const colorInput = create(0xff0000);
    expect(colorInput.shadowRoot?.innerHTML).toMatchSnapshot();
  });

  it('responds to user color changes', async () => {
    const colorInput = create(0xff0000);
    const changeHandler = vi.fn();
    colorInput.addEventListener('change', changeHandler);

    const colorPicker = getByDisplayValue(colorInput.shadowRoot as any, '#ff0000');

    // Simulate user selecting a new color - for color inputs we need to manually set value and fire change
    // @ts-ignore - tests
    colorPicker.value = '#00ff00';
    colorPicker.dispatchEvent(new Event('change', { bubbles: true }));

    expect(changeHandler).toHaveBeenCalledOnce();
    expect(changeHandler.mock.calls[0][0].detail.value).toBe(0x00ff00);
  });

  it('updates when color is programmatically set', () => {
    const colorInput = create(0xff0000);
    colorInput.setValue(0x0000ff);

    expect(colorInput.getValue()).toBe(0x0000ff);

    const colorPicker = getByDisplayValue(colorInput.shadowRoot as any, '#0000ff');
    expect(colorPicker).toBeTruthy();
  });
});
