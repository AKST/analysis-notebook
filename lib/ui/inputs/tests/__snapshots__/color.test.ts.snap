// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ColorInput > renders color input correctly 1`] = `
"<style>:host {
  display: block;
}

.color-input {
  display: block;
  width: 100%;
  height: 24px;
  outline: none;
  background: transparent;
  cursor: pointer;
  border-radius: 0;
  padding: 0;
  border: var(--border-outset-grey);
  &::-webkit-color-swatch {
    border: none;
  }

}

</style><input style="background-color: rgb(255, 0, 0);" type="color" class="color-input" value="#ff0000">"
`;
